include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,223 @@
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef DHT_STORAGE_HPP
+#define DHT_STORAGE_HPP
+
+#include <algorithm>
+#include <map>
+#include <set>
+#include <string>
+
+#include <libtorrent/config.hpp>
+#include <libtorrent/time.hpp>
+#include <libtorrent/socket.hpp>
+#include <libtorrent/sha1_hash.hpp>
+#include <libtorrent/bloom_filter.hpp>
+#include <libtorrent/address.hpp>
+#include <libtorrent/session_settings.hpp>
+#include <libtorrent/performance_counters.hpp>
+
+#include <libtorrent/kademlia/item.hpp>
+#include <libtorrent/kademlia/node_id.hpp>
+
+namespace libtorrent { namespace dht
+{
+	// this is the entry for every peer
+	// the timestamp is there to make it possible
+	// to remove stale peers
+	struct peer_entry
+	{
+		time_point added;
+		tcp::endpoint addr;
+		bool seed;
+	};
+
+	// this is a group. It contains a set of group members
+	struct torrent_entry
+	{
+		std::string name;
+		std::set<peer_entry> peers;
+	};
+
+	struct dht_immutable_item
+	{
+		dht_immutable_item() : value(0), num_announcers(0), size(0) {}
+		// malloced space for the actual value
+		char* value;
+		// this counts the number of IPs we have seen
+		// announcing this item, this is used to determine
+		// popularity if we reach the limit of items to store
+		bloom_filter<128> ips;
+		// the last time we heard about this
+		time_point last_seen;
+		// number of IPs in the bloom filter
+		int num_announcers;
+		// size of malloced space pointed to by value
+		int size;
+	};
+
+	struct ed25519_public_key { char bytes[item_pk_len]; };
+
+	struct dht_mutable_item : dht_immutable_item
+	{
+		char sig[item_sig_len];
+		boost::uint64_t seq;
+		ed25519_public_key key;
+		char* salt;
+		int salt_size;
+	};
+
+	struct count_peers
```

Comment:
I imagine this helper class as well as the comparator below are only used internally. if so, they should live in a .cpp file in the anonymous namespace